### ISSUE TITLE
Wrong signature inferred for path dependent type parameter

### ISSUE THREAD

**Issue by bishabosha** [lampepfl/dotty#8991]
## Minimized code

```Scala
object Macros {

  abstract class Universe {
    class Tree
    trait WeakTypeTag[T]
    object WeakTypeTag {
      implicit val IntTag: WeakTypeTag[Int] = ???
    }
  }

  trait Context {
    val universe: Universe
  }

  class MacroImpl(val c: Context) {
    import c.universe._
    def poly[T: WeakTypeTag]: Tree = ???
  }

  def test1 = new MacroImpl(???).poly[Int]
}

```

## Output

```scala
-- Error: MacroBundleTasty.scala:20:42 
20 |  def test1 = new MacroImpl(???).poly[Int]
   |                                          ^
   |no implicit argument of type Nothing was found for an implicit parameter of method poly in class MacroImpl
1 error found
```

## Expectation
Compile as in Scala 2.

The alternative below compiles, but the signature in TASTy for `MacroImpl.poly` is `poly[Signed Signature(List(1, scala.Nothing),Macros$.Universe.Tree)]`, which is incorrect
```Scala
object Macros {

  abstract class Universe {
    class Tree
    trait WeakTypeTag[T]
  }

  trait Context {
    val universe: Universe
  }

  class MacroImpl(val c: Context) {
    import c.universe._
    def poly[T: WeakTypeTag]: Tree = ???
  }

  def test1 = new MacroImpl(???).poly[Int](???)
}

```

**Comment by smarter** [comment]
It looks like we don't handle unstable prefixes like in Scala 2:
```scala
class Foo {
  class A
  object A {
    implicit val implicitA: A = new A
  }
  def get(implicit a: A): A = a
}

object Test {
  (new Foo).get
}
```
Dotty fails to find the implicit for `(new Foo).get`, but scalac figures out by adding a synthetic definition for the prefix to get a stable path:
```scala
  <synthetic> <stable> <artifact> val stabilizer$1: Foo = new Foo();
  stabilizer$1.get(stabilizer$1.A.implicitA)
```

> The alternative below compiles, but the signature in TASTy for MacroImpl.poly is poly[Signed Signature(List(1, scala.Nothing),Macros$.Universe.Tree)], which is incorrect

This sort of makes sense: the type of the parameter is ` c.universe.WeakTypeTag[T]` with `c` replaced by a stable prefix, but in the expression `new MacroImpl(???).poly[Int](???)` we don't have a stable prefix, so there's no way to represent this type (and no value we could possibly pass to this type), so we might as well represent it as `Nothing`.  A synthetic stable prefix would also fix this. 

In Dotty, synthetic definitions like this are normally created using a "lifter", see https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/typer/EtaExpansion.scala, perhaps we should add some logic to lift the prefix of an implicit method call when needed (which seems to be at a minimum when the prefix is unstable and the type is a class type).

Note that this can be worked around by manually using a stable prefix:
```scala
  def test1 = {
    val m = new MacroImpl(???)
    m.poly[Int]
  }
```

**Comment by smarter** [comment]
We should check in which situations exactly scalac introduces these stabilizer definitions. Note that some of these situations are already handle by skolemization which is much cheaper: https://github.com/lampepfl/dotty/commit/502c7c9fd700ad449cf8383cf75153999ee90ba9

**Comment by bishabosha** [comment]
>>The alternative below compiles, but the signature in TASTy for MacroImpl.poly is poly[Signed Signature(List(1, scala.Nothing),Macros$.Universe.Tree)], which is incorrect

>This sort of makes sense:
I thought signatures always mapped to the jvm signature of the symbol, but this appears to not be the case, a callsites' arguments should not affect it surely?

**Comment by smarter** [comment]
> My intuition was that signatures always mapped to the JVM signature of a matching overload,  but this appears to not be the case

Yeah it's a bt confusing: Dotty uses signatures to distinguish between overloads and overrides, so signatures are always computed as [_seen from_](https://www.scala-lang.org/files/archive/spec/2.13/03-types.html#base-types-and-member-definitions) some prefix, e.g. in:
```scala
class Foo[A] {
  def foo(x: A): String = "foo"
}
class Bar extends Foo[Int] {
  override def foo(x: A): String = "bar"
}
```
There's two definitions for `foo` in `Bar`, so we find two different denotations, but their signatures as seen from an instance of `Bar` is the same: `Signature(List(scala.Int), java.lang.String`, so their denotations get [merged together ](https://github.com/lampepfl/dotty/blob/76b42dd1b90156472d5d31b237f5ca587e1f6e9e/compiler/src/dotty/tools/dotc/core/Denotations.scala#L106) (the signatures don't have to exactly match, the result type [may be different](https://github.com/lampepfl/dotty/blob/76b42dd1b90156472d5d31b237f5ca587e1f6e9e/compiler/src/dotty/tools/dotc/core/Denotations.scala#L1126)).
